### PR TITLE
BUGZ-811: Fix bugs in EntityTreeRenderer::updateChangedEntities

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.atlassian.net/browse/BUGZ-811

### CODE REVIEW

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -207,7 +207,7 @@ void EntityTreeRenderer::stopDomainAndNonOwnedEntities() {
 
             if (entityItem && !entityItem->getScript().isEmpty()) {
                 if (!(entityItem->isLocalEntity() || (entityItem->isAvatarEntity() && entityItem->getOwningAvatarID() == getTree()->getMyAvatarSessionUUID()))) {
-                    if (entityItem->contains(_avatarPosition)) {
+                    if (_currentEntitiesInside.contains(entityID)) {
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: I changed this in a few places.  This check is just faster (and slightly more accurate).

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 2]
```diff
@@ -389,13 +391,7 @@ void EntityTreeRenderer::updateChangedEntities(const render::ScenePointer& scene
     PerformanceTimer pt("change");
     std::unordered_set<EntityItemID> changedEntities;
     _changedEntitiesGuard.withWriteLock([&] {
-#if 0
-        // FIXME Weird build failure in latest VC update that fails to compile when using std::swap
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: idk what was up with this but it seems fine now, I can revert if it causes problems

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 3]
```diff
@@ -450,43 +445,40 @@ void EntityTreeRenderer::updateChangedEntities(const render::ScenePointer& scene
...
         }
         {
-            PROFILE_RANGE_EX(simulation_physics, "UpdateRenderables", 0xffff00ff, sortedRenderables.size());
+            PROFILE_RANGE_EX(simulation_physics, "SortAndUpdateRenderables", 0xffff00ff, sortedRenderables.size());
 
             // compute remaining time budget
+            const auto& sortedRenderablesVector = sortedRenderables.getSortedVector();
             uint64_t updateStart = usecTimestampNow();
-            uint64_t timeBudget = MIN_SORTED_UPDATE_RENDERABLES_TIME_BUDGET;
             uint64_t sortCost = updateStart - sortStart;
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: sortCost needs to be computed below the call to `getSortedVector`

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 4]
```diff
@@ -450,43 +445,40 @@ void EntityTreeRenderer::updateChangedEntities(const render::ScenePointer& scene
...
                 }
                 const auto& renderable = sortedRenderable.getRenderer();
                 renderable->updateInScene(scene, transaction);
-                _renderablesToUpdate.erase(renderable->getEntity()->getID());
+                _renderablesToUpdate.erase(renderable);
             }
 
             // compute average per-renderable update cost
-            size_t numUpdated = numSorted - sortedRenderables.size() + 1; // add one to avoid divide by zero
+            size_t numUpdated = sortedRenderables.size() - _renderablesToUpdate.size() + 1; // add one to avoid divide by zero
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: Before this was always numSorted - sortedRenderables.size() where numSorted = sortedRenderables.size(), so we always thought we'd only updated 1 thing

File: libraries/shared/src/PrioritySortUtil.h [hunk 5]
```diff
@@ -45,7 +45,7 @@ namespace PrioritySortUtil {
     class PriorityQueue {
     public:
         PriorityQueue() = delete;
-        PriorityQueue(const ConicalViewFrustums& views) : _views(views) { }
+        PriorityQueue(const ConicalViewFrustums& views) : _views(views), _usecCurrentTime(usecTimestampNow()) { }
```
Comment (SamGondelman) [inline libraries/shared/src/PrioritySortUtil.h]: This wasn't being set before.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20119/HighFidelity-Beta-PR15834-c605c29.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20119/HighFidelity-Quest-Beta-PR15834-c605c29.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20119/HighFidelity-Beta-PR15834-c605c29.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20119/HighFidelity-Beta-Interface-PR15834-c605c29.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20119/HighFidelity-Beta-PR15834-c605c29.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20119/HighFidelity-Beta-Interface-PR15834-c605c29.dmg

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 6]
```diff
@@ -450,43 +445,40 @@ void EntityTreeRenderer::updateChangedEntities(const render::ScenePointer& scene
         PrioritySortUtil::PriorityQueue<SortableRenderer> sortedRenderables(views);
         sortedRenderables.reserve(_renderablesToUpdate.size());
         {
-            PROFILE_RANGE_EX(simulation_physics, "SortRenderables", 0xffff00ff, (uint64_t)_renderablesToUpdate.size());
-            std::unordered_map<EntityItemID, EntityRendererPointer>::iterator itr = _renderablesToUpdate.begin();
-            while (itr != _renderablesToUpdate.end()) {
-                assert(itr->second); // only valid renderables are added to _renderablesToUpdate
-                sortedRenderables.push(SortableRenderer(itr->second));
-                ++itr;
+            PROFILE_RANGE_EX(simulation_physics, "BuildSortedRenderables", 0xffff00ff, (uint64_t)_renderablesToUpdate.size());
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: The profile category is wrong here.  It shouldn't be `simulation_physics` but something else more to do with `render`.  However, I see that there is much incorrect categorization going on in `EntityTreeRenderer`.  Maybe we can eventually fix it in some other PR.

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp (hunk 4 above)
Comment (AndrewMeadows) [reply libraries/entities-renderer/src/EntityTreeRenderer.cpp]: Whoops.  Thanks for fixing.

**Pull request merged**